library/java/org/cronvoy/BidirectionalStreamBuilderImpl.java

```diff
@@ -0,0 +1,145 @@
+
+  private boolean delayRequestHeadersUntilFirstFlush;
+
+  // Request reporting annotations.
+  private Collection<Object> requestAnnotations;
+
+  private boolean trafficStatsTagSet;
+  private int trafficStatsTag;
+  private boolean trafficStatsUidSet;
+  private int trafficStatsUid;
+
+  /**
+   * Creates a builder for {@link BidirectionalStream} objects. All callbacks for generated {@code
+   * BidirectionalStream} objects will be invoked on {@code executor}. {@code executor} must not run
+   * tasks on the current thread, otherwise the networking operations may block and exceptions may
+   * be thrown at shutdown time.
+   *
+   * @param url the URL for the generated stream
+   * @param callback the {@link BidirectionalStream.Callback} object that gets invoked upon
+   *     different events occuring
+   * @param executor the {@link Executor} on which {@code callback} methods will be invoked
+   * @param cronetEngine the {@link CronetEngine} used to create the stream
+   */
+  BidirectionalStreamBuilderImpl(String url, BidirectionalStream.Callback callback,
+                                 Executor executor, CronvoyEngineBase cronetEngine) {
+    super();
+    if (url == null) {
+      throw new NullPointerException("URL is required.");
+    }
+    if (callback == null) {
+      throw new NullPointerException("Callback is required.");
+    }
+    if (executor == null) {
+      throw new NullPointerException("Executor is required.");
+    }
+    if (cronetEngine == null) {
+      throw new NullPointerException("CronetEngine is required.");
+    }
+    this.url = url;
+    this.callback = callback;
+    this.executor = executor;
+    this.cronetEngine = cronetEngine;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl setHttpMethod(String method) {
+    if (method == null) {
+      throw new NullPointerException("Method is required.");
+    }
+    httpMethod = method;
+    return this;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl addHeader(String header, String value) {
+    if (header == null) {
+      throw new NullPointerException("Invalid header name.");
+    }
+    if (value == null) {
+      throw new NullPointerException("Invalid header value.");
+    }
+    requestHeaders.add(new AbstractMap.SimpleImmutableEntry<>(header, value));
+    return this;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl setPriority(@StreamPriority int priority) {
+    this.priority = priority;
+    return this;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl
+  delayRequestHeadersUntilFirstFlush(boolean delayRequestHeadersUntilFirstFlush) {
+    this.delayRequestHeadersUntilFirstFlush = delayRequestHeadersUntilFirstFlush;
+    return this;
+  }
+
+  @Override
+  public ExperimentalBidirectionalStream.Builder addRequestAnnotation(Object annotation) {
+    if (annotation == null) {
+      throw new NullPointerException("Invalid metrics annotation.");
+    }
+    if (requestAnnotations == null) {
```

Comment:
Could we just initialize `requestAnnotations` as an empty list?

Reply:
This class is not yet used - Bidirectional Streams are not yet supported by Cronvoy. It will likely go through much more adaptation. The only reason why I kept it was to have the whole "skeleton" exposed.

Reply:
Oh what I meant to say was if we could initialize it to an empty list and avoid a null check here (since we are just checking and assigning it here anyway)

Reply:
The original code does expect this object to be null when no request annotations are provided:

https://chromium.googlesource.com/chromium/src/+/main/components/cronet/android/java/src/org/chromium/net/impl/RequestFinishedInfoImpl.java#56

I don't know why they did that. Strange optimization. But at this point in time, I am not yet ready to start fixing this.